From 25edc144e718ce14313aa16c3538347cc6b42114 Mon Sep 17 00:00:00 2001
From: Yannick Fertre <yannick.fertre@foss.st.com>
Date: Thu, 27 Jun 2024 15:45:44 +0200
Subject: [PATCH 1031/1695] plat-stm32mp2: display: add wrapper

Add a display wrapper to register all secure devices which
need to be initialized during OPTEE init sequence.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 134e6178d2905437bf98ef728bc2efd71e8ee003 plat-stm32mp2: display: add wrapper]
Change-Id: I6eefc320fb326878a60d82e406d50f6c23a392d5
Signed-off-by: Yannick Fertre <yannick.fertre@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/391784
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
---
 core/arch/arm/plat-stm32mp2/display.c | 170 ++++++++++++++++++++++++++
 core/arch/arm/plat-stm32mp2/sub.mk    |   2 +
 2 files changed, 172 insertions(+)
 create mode 100644 core/arch/arm/plat-stm32mp2/display.c

diff --git a/core/arch/arm/plat-stm32mp2/display.c b/core/arch/arm/plat-stm32mp2/display.c
new file mode 100644
index 000000000..204fe018d
--- /dev/null
+++ b/core/arch/arm/plat-stm32mp2/display.c
@@ -0,0 +1,170 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2020-2024, STMicroelectronics
+ * Copyright (c) 2016, Linaro Limited
+ */
+
+#include <assert.h>
+#include <display.h>
+#include <drivers/frame_buffer.h>
+#include <io.h>
+#include <kernel/boot.h>
+#include <kernel/delay.h>
+#include <kernel/dt.h>
+#include <kernel/panic.h>
+#include <libfdt.h>
+#include <mm/core_mmu.h>
+#include <mm/core_memprot.h>
+#include <stddef.h>
+#include <stdint.h>
+#include <tee_api_types.h>
+#include <trace.h>
+
+static SLIST_HEAD(, disp_dev_list) display_device_list =
+	SLIST_HEAD_INITIALIZER(disp_dev_list);
+
+/*
+ * Trusted UI frame buffer description
+ */
+static struct frame_buffer tui_frame_buffer = {
+	.width      = 480,
+	.height     = 272,
+	.width_dpi  = 160,
+	.height_dpi = 160,
+	.bpp        = FB_24BPP, /* never used actually.*/
+};
+
+struct frame_buffer *display_get_frame_buffer(void)
+{
+	return &tui_frame_buffer;
+}
+
+TEE_Result display_get_fb_addr_from_dtb(paddr_t *pa, size_t *size)
+{
+	const void *fdt = get_embedded_dt();
+	int node = fdt_path_offset(fdt, "/reserved-memory");
+	paddr_t fb_pa = 0;
+	size_t fb_sz = 0;
+
+	assert(pa && size);
+
+	if (node < 0) {
+		EMSG("Can't find reserved memory");
+		return TEE_ERROR_ITEM_NOT_FOUND;
+	}
+
+	/* Note physical address 0x0 is reserved in OP-TEE core */
+	node = fdt_subnode_offset(fdt, node, "ltdc-sec-layer");
+	if (node < 0) {
+		EMSG("Can't find ltdc-sec-layer");
+		return TEE_ERROR_ITEM_NOT_FOUND;
+	}
+
+	fb_pa = fdt_reg_base_address(fdt, node);
+	fb_sz = fdt_reg_size(fdt, node);
+
+	DMSG("TUI frame buffer phys range: [%#"PRIxPA" %#"PRIxPA"]",
+	     fb_pa, fb_pa + fb_sz - 1);
+
+	if (fb_pa == DT_INFO_INVALID_REG ||
+	    fb_sz > CFG_TUI_FRAME_BUFFER_SIZE_MAX) {
+		DMSG("Can't retrieve TUI frame buffer location");
+		return TEE_ERROR_GENERIC;
+	}
+
+	*pa = fb_pa;
+	*size = fb_sz;
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result get_fb_from_dtb(struct frame_buffer *fb)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	paddr_t fb_pa = 0;
+	size_t fb_sz = 0;
+
+	if (!fb)
+		return TEE_ERROR_GENERIC;
+
+	res = display_get_fb_addr_from_dtb(&fb_pa, &fb_sz);
+	if (res)
+		return res;
+
+	if (!core_mmu_add_mapping(MEM_AREA_RAM_SEC, fb_pa, fb_sz)) {
+		EMSG("Can't map TUI frame buffer");
+		return TEE_ERROR_GENERIC;
+	}
+
+	fb->base = phys_to_virt(fb_pa, MEM_AREA_RAM_SEC, fb_sz);
+	assert(fb->base);
+
+	DMSG("TUI frame buffer virt range: [%p %#"PRIxPTR"]",
+	     fb->base, (uintptr_t)fb->base + fb_sz - 1);
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result display_init(void)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	uint32_t width = 0;
+	uint32_t height = 0;
+	struct disp_dev_list *ddev = NULL;
+
+	if (get_fb_from_dtb(&tui_frame_buffer))
+		panic();
+
+	SLIST_FOREACH(ddev, &display_device_list, link) {
+		if (ddev->device_init) {
+			res = ddev->device_init(ddev->device);
+			if (res != TEE_SUCCESS)
+				return res;
+		}
+	}
+
+	SLIST_FOREACH(ddev, &display_device_list, link) {
+		if (ddev->device_get_display_size) {
+			res = ddev->device_get_display_size(ddev->device,
+							    &width,
+							    &height);
+			if (res != TEE_SUCCESS)
+				return res;
+		}
+	}
+
+	DMSG("TUI frame buffer size: %"PRIu32" x %"PRIu32, width, height);
+	tui_frame_buffer.width = width;
+	tui_frame_buffer.height = height;
+
+	SLIST_FOREACH(ddev, &display_device_list, link) {
+		if (ddev->device_activate) {
+			res = ddev->device_activate(ddev->device,
+						    display_get_frame_buffer(),
+						    0, 0);
+			if (res != TEE_SUCCESS)
+				return res;
+		}
+	}
+
+	return TEE_SUCCESS;
+}
+
+void display_final(void)
+{
+	struct disp_dev_list *ddev = NULL;
+	TEE_Result res = TEE_ERROR_GENERIC;
+
+	SLIST_FOREACH(ddev, &display_device_list, link) {
+		if (ddev->device_final) {
+			res = ddev->device_final(ddev->device);
+			if (res != TEE_SUCCESS)
+				return;
+		}
+	}
+}
+
+void display_register_device(struct disp_dev_list *ddev)
+{
+	SLIST_INSERT_HEAD(&display_device_list, ddev, link);
+}
diff --git a/core/arch/arm/plat-stm32mp2/sub.mk b/core/arch/arm/plat-stm32mp2/sub.mk
index c524045d9..7752720bf 100644
--- a/core/arch/arm/plat-stm32mp2/sub.mk
+++ b/core/arch/arm/plat-stm32mp2/sub.mk
@@ -3,5 +3,7 @@ global-incdirs-y += .
 srcs-y += main.c
 srcs-y += stm32mp_pm.c
 subdirs-y += drivers
+
+srcs-$(CFG_DISPLAY) += display.c
 srcs-$(CFG_SCMI_CORTEXM_AGENT) += scmi_cortexm_agent.c
 srcs-$(CFG_SCMI_SCPFW) += scmi_server_scpfw.c
-- 
2.47.2

